### ISSUE TITLE
Allow for typesafe extension of the AxiosConfiguration

### ISSUE THREAD

**Issue by mtandersson** [axios/axios#1963]
#### Summary

Allow for typesafe extension of AxiosConfiguration

#### Context

Allow for the following user case (without the cast)
```
    interface Config extends AxiosRequestConfig {
      start?: Date
    }
    const requestConfig: Config = {
      ...
    }
    const instance = axios.create(requestConfig)

    instance.interceptors.request.use(axiosConfig => {
      const config = axiosConfig as Config // Cast is needed as axiosConfig is infered to AxiosConfiguration and not Config
      config.start = new Date()
      return config
    })

```

**Comment by bailer** [comment]
I am also in a similar position. I have an interceptor that adds a header if a specific field is present in the configuration, a field which is not part of the original AxiosRequestConfig but would be nice if I could extend it for the instance. My suggestion would be to add a generic argument to the create function.

**Comment by jasonsaayman** [comment]
Closed merged #1964
